### ISSUE TITLE
💡 RFC: Expose slots to components

### ISSUE THREAD

**Issue by jonathantneal** [snowpackjs/astro#1301]
### Background & Motivation

When passing content to a component via slots, sometimes there is a need to conditionally render a wrapping element or class based on the presence or absence of content in a slot.

For example, you might want to display a subheading **only** when the `subtitle` slot is provided:

```astro
---
<card-component class="card">
  <h3>
    <slot name="title" />
  </h3>
  <!-- only display when subtitle is provided -->
  {
    Astro.slots.subtitle && (
      <h4>
        <slot name="subtitle" />
      </h4>
    )
  }
</card-component>
```

There is currently no way to programmatically access content in a slot, so it is not possible to conditionally display an element when a specific slot has content.

This use case only requires knowledge of whether there is or is not content for a given slot. Other use cases might require the content itself to be exposed.

Other frameworks expose some information about slots for this purpose.

- [Svelte: Check for slot content](https://svelte.dev/tutorial/optional-slots)
- [Vue: Determining if slot content is null or empty](https://stackoverflow.com/questions/47432702/determining-if-slot-content-is-null-or-empty/47434345)

### Proposed Solution

Provide the ability to access slotted content via `Astro.slots`.

```astro
---
<card-component class="card">
  <h3>
    <slot name="title" />
  </h3>
  <!-- only display when subtitle is provided -->
  {
    Astro.slots.subtitle && (
      <h4>
        <slot name="subtitle" />
      </h4>
    )
  }
</card-component>
```

The default slot should be accessible via `Astro.slots.default`.

If no content has been provided for a slot, `Astro.slots` will not contain a property for that slot.

### Alternatives considered

Content could be passed via properties (`Astro.props`) instead of slots, but then we might lose the power and convenience of rendering content in slots.

An additional property could be passed to indicate whether a slot contains content or not, but this isn't _DRY_ and so it may be error-prone.

### Open Questions

> Is it enough to know that there is or is not content for a given slot, or would there have to be some way to manipulate content before it's rendered?

> Should it be possible to programmatically _add_ and _remove_ properties to `Astro.slots` in order to change what is rendered?

### Additional credits

I could not have shared this RFC without the help and co-authoring of @openidauthority. Thank you!

### Detailed Design

Implemented here: https://github.com/snowpackjs/astro/pull/1225

### Help make it happen!

- [X] I am willing to submit a PR to implement this change.
- [ ] I am willing to submit a PR to implement this change, but would need some guidance.
- [ ] I am not willing to submit a PR to implement this change.

**Comment by jasikpark** [comment]
title please! 😆

**Comment by jonathantneal** [comment]
Of all the things to forget. Done. Thank you, @jasikpark!

**Comment by jasikpark** [comment]
I'd say this would be great to add to the next RFC meeting on Sept 14th: #1278!

**Comment by drwpow** [comment]
RFC call notes:
- Implement as boolean to start; easier to change later
- Edge cases (if any) can be evaluated later (passing in Fragments as slots, etc.)

**Comment by jonathantneal** [comment]
Thanks! I hope this can now be fulfilled in https://github.com/snowpackjs/astro/pull/1368

**Comment by matthewp** [comment]
Closed in #1368

**Comment by jasikpark** [comment]
I'm interested in seeing access to the content of slots, in addition to allowing the detection of filled slots that we have now, but this issue has definitely been completed in the RFC meeting priorities!
